### PR TITLE
EVENTLISTS [readevent]

### PR DESCRIPTION
**Pull request by kashish2210**
This PR includes the update to events.JL function so that it contains now abstract methods with a constructor, and mainly comprises the light curve function with metadata and filter as per the discussion
majorly work on this #29

lightcurve function included create_lightcurve,EventProperty, AbstractLightCurve ,rebin, calculate_errors, LightCurve, LightCurveMetadata

tagging @matteobachetti  @fjebaker  @stefanocovino

### CODE REVIEW

File: src/events.jl [hunk 1]
```diff
@@ -60,26 +74,60 @@ function readevents(path; T = Float64)
...
+                if isempty(times) && ("TIME" in colnames)
                     times = convert(Vector{T}, read(hdu, "TIME"))
                 end
-                if "ENERGY" in colnames
+                if isempty(energies) && ("ENERGY" in colnames)
                     energies = convert(Vector{T}, read(hdu, "ENERGY"))
                 end
+
+                # If we found both time and energy data, we can return
+                if !isempty(times) && !isempty(energies)
```
Comment (matteobachetti) [inline src/events.jl:88]: Note that most event lists do _not_ have an "energy" column, so this will keep going through the different extension needlessly. 
When you find the time column, you will also have found all the information you can find about single events (energy, PI, etc.).

File: src/lightcurve.jl [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+"""
+function calculate_errors(
+    counts::Vector{Int},
+    method::Symbol,
+    exposure::Vector{T},
+) where {T}
+    if method === :poisson
+        return convert.(T, sqrt.(counts))
+    elseif method === :gaussian
+        return convert.(T, sqrt.(counts .+ 1))
```
Comment (matteobachetti) [inline src/lightcurve.jl:63]: Gaussian errors need to be given by the user, estimating them like this would be wrong and, in any case, it seems to assume they are Poisson (the sqrt only makes sense in that case, and summing 1 is a quick way to give an error bar when the count rate is 0).

**Review by fjebaker — COMMENTED** [review]
This looks generally really good! I think there are just some redundancies and simplifications that ought to be made to help us maintain this code in the future.

We should also talk about tests and how tests should be organised. We can chat about that in the meeting tomorrow. I'd recommend not being too specific with the tests at this stage, else you'll spend most of your time maintaining the tests when you need to make a small change to how something works.

**Review by fjebaker — COMMENTED** [review]
Thanks @kashish2210 this looks fine! Just some small things remaining. The code is getting quite complex, so I'd urge you to split things up into small, reusable functions in future work, but we can clean this up later in different PRs.

**Comment by kashish2210** [comment]
@fjebaker I have made the changes Here is the summary  
## Specific catch blocks: Replaced broad catch e with specific error handling:
```
catch header_error for header reading
catch energy_error for energy column reading
catch sector_error for sector column reading
catch hdu_error for HDU access errors
catch table_error for table reading errors

```
## Reduced nesting: Simplified the logic flow by:
```
Using findfirst() instead of loops for column searching
Extracting the main HDU reading logic to reduce nesting
Using clearer conditional structures
```

**Comment by kashish2210** [comment]
@fjebaker, I have added only events hereshallmakediffrent PR for lightcurve cause this PR getting bit complex i guess

File: src/events.jl [hunk 3]
```diff
@@ -1,85 +1,694 @@
...
+        time_col = findfirst(col -> uppercase(col) == "TIME", all_cols)
+        
+        if isnothing(time_col)
+            error("TIME column not found in HDU $hdu")
+        end
+        
+        actual_time_col = all_cols[time_col]
+        
+        # Read time column with case-insensitive option
+        time = convert(Vector{T}, read(selected_hdu, actual_time_col, case_sensitive=false))
```
Comment (fjebaker) [inline src/events.jl:591]:
```suggestion
        time = convert(Vector{T}, read(selected_hdu, "TIME", case_sensitive=false))
```

What's wrong with this? It will raise an error if no time column is found already, and is less code to reason about.

File: test/test_events.jl [hunk 4]
```diff
@@ -1,149 +1,453 @@
...
-                    @test length(data.metadata.headers) > 0
-                finally
-                    global_logger(old_logger)
-                end
-            end
-        else
-            @info "Test file '$(test_filepath)' not found. Skipping this test."
-        end
-    end
+    ev = EventList(times, energies)
```
Comment (fjebaker) [inline test/test_events.jl:33]: Nice (~;

File: test/test_events.jl [hunk 5]
```diff
@@ -1,149 +1,453 @@
...
+    @test length(ev) == 5
+    @test has_energies(ev)
+
+    # Test with no energies
+    ev_no_energy = EventList(times)
+    @test ev_no_energy.times == times
+    @test isnothing(ev_no_energy.energies)
+    @test !has_energies(ev_no_energy)
+
+    println("✓ Basic EventList creation tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:45]:
```suggestion
```
Not needed. The Test.jl library already provides all this information.

File: test/test_events.jl [hunk 6]
```diff
@@ -1,149 +1,453 @@
...
+
+    # Test energies() accessor  
+    @test energies(ev) === ev.energies
+    @test energies(ev) == energies_vec
+
+    # Test with nothing energies
+    ev_no_energy = EventList(times_vec)
+    @test isnothing(energies(ev_no_energy))
+
+    println("✓ Accessor function tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:67]:
```suggestion
```

File: test/test_events.jl [hunk 7]
```diff
@@ -1,149 +1,453 @@
...
+
+    # Test length
+    @test length(ev) == 4
+    @test length(ev) == length(times_vec)
+
+    # Test size
+    @test size(ev) == (4,)
+    @test size(ev) == (length(times_vec),)
+
+    println("✓ Base interface method tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:84]:
```suggestion
```

File: test/test_events.jl [hunk 8]
```diff
@@ -1,149 +1,453 @@
...
+        Dict{String,Any}(),
+    )
+    ev_extra = EventList(times_extra, energies_extra, dummy_meta)
+
+    filter_time!(t -> t >= 2.5, ev_extra)
+    @test ev_extra.times == [3.0, 4.0]
+    @test ev_extra.energies == [30.0, 40.0]
+    @test ev_extra.meta.extra_columns["INDEX"] == [3, 4]
+
+    println("✓ filter_time! function tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:131]:
```suggestion
```

File: test/test_events.jl [hunk 9]
```diff
@@ -1,149 +1,453 @@
...
+        Dict{String,Any}(),
+    )
+    ev_extra = EventList(times_extra, energies_extra, dummy_meta)
+
+    filter_energy!(e -> e >= 30.0, ev_extra)
+    @test ev_extra.times == [3.0, 4.0]
+    @test ev_extra.energies == [35.0, 45.0]
+    @test ev_extra.meta.extra_columns["DETX"] == [0.3, 0.4]
+
+    println("✓ filter_energy! function tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:175]:
```suggestion
```

File: test/test_events.jl [hunk 10]
```diff
@@ -1,149 +1,453 @@
...
+    @test ev_extra.energies == [10.0, 30.0, 50.0]
+    @test ev_extra.meta.extra_columns["FLAG"] == [1, 1, 1]
 
-        # Test with invalid FITS file
-        invalid_file = tempname()
-        open(invalid_file, "w") do io
-            write(io, "This is not a FITS file")
-        end
-        @test_throws Exception readevents(invalid_file)
+    println("✓ filter_on! function tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:225]:
```suggestion
```

File: test/test_events.jl [hunk 11]
```diff
@@ -1,149 +1,453 @@
...
+    ev_filtered_no_energy = filter_time(t -> t > 2.0, ev_no_energy)
+
+    @test ev_no_energy.times == [1.0, 2.0, 3.0, 4.0]  # Original unchanged
+    @test ev_filtered_no_energy.times == [3.0, 4.0]
+    @test isnothing(ev_filtered_no_energy.energies)
+
+    # Test filter_energy error with no energies
+    @test_throws AssertionError filter_energy(e -> e > 10.0, ev_no_energy)
+
+    println("✓ Non-mutating filter function tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:271]:
```suggestion
```

File: test/test_events.jl [hunk 12]
```diff
@@ -1,149 +1,453 @@
...
+    # Test edge case: no filtering needed (all pass)
+    ev_all_pass = EventList([1.0, 2.0, 3.0], [10.0, 20.0, 30.0])
+    original_times = copy(ev_all_pass.times)
+    original_energies = copy(ev_all_pass.energies)
+    filter_time!(t -> t > 0.0, ev_all_pass)  # All should pass
+    @test ev_all_pass.times == original_times
+    @test ev_all_pass.energies == original_energies
+    @test length(ev_all_pass) == 3
+
+    println("✓ Complex filtering scenario tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:310]:
```suggestion
```

File: test/test_events.jl [hunk 13]
```diff
@@ -1,149 +1,453 @@
...
+    @test length(data.meta.headers) >= 1  # Should have at least primary header
+    @test length(data.meta.extra_columns) == 0
+
+    # Test reading with extra columns
+    data = readevents(sample; extra_columns = ["INDEX"])
+    @test length(data.meta.extra_columns) == 1
+    @test haskey(data.meta.extra_columns, "INDEX")
+    @test data.meta.extra_columns["INDEX"] == collect(1:5)
+
+    println("✓ Basic readevents tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:332]:
```suggestion
```

File: test/test_events.jl [hunk 14]
```diff
@@ -1,149 +1,453 @@
...
+    event_table["ENERGY"] = energies
+    write(f, event_table)
+    close(f)
+
+    # Test specifying specific HDU
+    data_hdu3 = readevents(sample_file, hdu = 3)
+    @test data_hdu3.times == times
+    @test data_hdu3.energies == energies
+
+    println("✓ HDU handling tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:362]:
```suggestion
```

File: test/test_events.jl [hunk 15]
```diff
@@ -1,149 +1,453 @@
...
+
+    # Test with PHA column
+    pha_file = mock_data(times, pi_values; energy_column = "PHA")
+
+    data_pha = readevents(pha_file)
+    @test data_pha.times == times
+    @test data_pha.energies == pi_values
+    @test data_pha.meta.energy_units == "PHA"
+
+    println("✓ Alternative energy column tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:386]:
```suggestion
```

File: test/test_events.jl [hunk 16]
```diff
@@ -1,149 +1,453 @@
...
+    table["TIME"] = times
+    write(f, table)
+    close(f)
+
+    data = readevents(time_only_file)
+    @test data.times == times
+    @test isnothing(data.energies)
+    @test isnothing(data.meta.energy_units)
+
+    println("✓ Missing column tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:408]:
```suggestion
```

File: test/test_events.jl [hunk 17]
```diff
@@ -1,149 +1,453 @@
...
+
+    # Test invalid FITS file
+    test_dir = mktempdir()
+    invalid_file = joinpath(test_dir, "invalid.fits")
+    open(invalid_file, "w") do io
+        write(io, "This is not a FITS file")
     end
+    @test_throws Exception readevents(invalid_file)
+
+    println("✓ Error handling tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:424]:
```suggestion
```

File: test/test_events.jl [hunk 18]
```diff
@@ -1,149 +1,453 @@
...
+    table["time"] = times  # lowercase
+    table["energy"] = energies  # lowercase
+    write(f, table)
+    close(f)
+
+    data = readevents(sample_file)
+    @test data.times == times
+    @test data.energies == energies
+
+    println("✓ Case insensitive column tests passed")
```
Comment (fjebaker) [inline test/test_events.jl:452]:
```suggestion
```

**Review by fjebaker — CHANGES_REQUESTED** [review]
This is great @kashish2210 and thanks for splitting the events list and lightcurve up so that we can review them in small chunks!

Don't add anything new to this PR. If you can just address the changes I've suggested, then from my end I'm happy for this to be merged!

**Comment by kashish2210** [comment]
Hey @matteobachetti Thank you for the feedback, I am replying your [comment](https://github.com/StingraySoftware/Stingray.jl/pull/51#discussion_r2135520170) about `function filter_energy(f, ev::EventList)` from #51 
## reason
`filter_energy!(f, ev::EventList)` - modifies the EventList in-place
`filter_energy(f, ev::EventList)` - returns a new filtered EventList (non-mutating)
This follows Julia's standard convention where functions ending with `!` are mutating (modify their arguments) and functions without `!` are non-mutating (return new objects). The same pattern applies to filter_time! vs filter_time.
However, I understand your suggestion about using a keyword argument instead. We could consolidate this into a single function like:
```julia
function filter_energy(f, ev::EventList; inplace::Bool = false)
    if inplace
        return filter_energy!(f, ev)
    else
        new_ev = deepcopy(ev)
        return filter_energy!(f, new_ev)
    end
end
```
@fjebaker @matteobachetti shall I implement this if you guys say i will add this in this PR

**Review by matteobachetti — APPROVED** [review]
Ok for me, I think it's a good starting point for the rest of the open PRs

**Review by fjebaker — APPROVED** [review]
Nice work! :+1:

**Comment by kashish2210** [comment]
thanks
